Pandas Data Frames
==========

The DataFrame is the way Pandas represents a table, and Series is the data-structure Pandas use to represent a column. So, a data frame and series are synonomous with table and column.

What makes Pandas so attractive is the powerful interface to access individual records of the table, proper handling of missing values, and relational-databases operations between DataFrames.

## Selecting values

Remember that a DataFrame provides a index as a way to identify the rows of the table. A row also has a position inside the table as well as a label, which uniquely identifies its entry in the DataFrame.

To access a value at the position [ i , j ] (row, column) of a DataFrame, we have two options, depending on what is the meaning of i in use.


First, print the first five rows of our data frame (`head` and `tails` are two functions which will show the first x number or last x number of rows).

In [53]:
import pandas
df = pandas.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')
print(df.head(10))

                        gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Albania                    1601.056136     1942.284244     2312.888958   
Austria                    6137.076492     8842.598030    10750.721110   
Belgium                    8343.105127     9714.960623    10991.206760   
Bosnia and Herzegovina      973.533195     1353.989176     1709.683679   
Bulgaria                   2444.286648     3008.670727     4254.337839   
Croatia                    3119.236520     4338.231617     5477.890018   
Czech Republic             6876.140250     8256.343918    10136.867130   
Denmark                    9692.385245    11099.659350    13583.313510   
Finland                    6424.519071     7545.415386     9371.842561   
France                     7029.809327     8662.834898    10560.485530   

                        gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                              

We can see that the first value in the first column is `1601.056136`. Remember that in programming, we begin our counting at zero, so this index would be `0,0`, not `1,1`.

## Use DataFrame.iloc[..., ...] to select values by their position
* Allows you to specify location by numerical index similar to 2D version of character selection in strings.


In [51]:
print("\nData value in first row at first column: ", data.iloc[0, 0])


Data value in first column of first row:  6137.076492


In [60]:
print("\nData value in fifth row at third column: ", data.iloc[4, 2])


Data value in fifth row at third column:  4254.337839


## Use `DataFrame.loc[..., ...]` to select values by their (entry) label.

*   Can specify location by name or by numerical index.

In [64]:
print(data.loc["Albania", "gdpPercap_1952"])

1601.056136


In [63]:
print(data.loc['Bulgaria', 'gdpPercap_1962'])

4254.337839


## Use `:` on its own to mean all columns or all rows.

*   Just like Python's usual slicing notation, we can print all columns or all rows with `.loc` using the `:`

In [74]:
print(data.loc["Albania",:])

gdpPercap_1952    1601.056136
gdpPercap_1957    1942.284244
gdpPercap_1962    2312.888958
gdpPercap_1967    2760.196931
gdpPercap_1972    3313.422188
gdpPercap_1977    3533.003910
gdpPercap_1982    3630.880722
gdpPercap_1987    3738.932735
gdpPercap_1992    2497.437901
gdpPercap_1997    3193.054604
gdpPercap_2002    4604.211737
gdpPercap_2007    5937.029526
Name: Albania, dtype: float64


* Would get the same result printing `data.iloc[0]` (without a second index).
* We can also omit the `:` and get the same result in either case.
    * e.g. `data.loc["Albania"]`

In [75]:
print(data.loc[:, "gdpPercap_1952"])

country
Albania                    1601.056136
Austria                    6137.076492
Belgium                    8343.105127
Bosnia and Herzegovina      973.533195
Bulgaria                   2444.286648
Croatia                    3119.236520
Czech Republic             6876.140250
Denmark                    9692.385245
Finland                    6424.519071
France                     7029.809327
Germany                    7144.114393
Greece                     3530.690067
Hungary                    5263.673816
Iceland                    7267.688428
Ireland                    5210.280328
Italy                      4931.404155
Montenegro                 2647.585601
Netherlands                8941.571858
Norway                    10095.421720
Poland                     4029.329699
Portugal                   3068.319867
Romania                    3144.613186
Serbia                     3581.459448
Slovak Republic            5074.659104
Slovenia                   4215.041741
Spain            

*   Would get the same result printing `data["gdpPercap_1952"]`
*   Also get the same result printing `data.gdpPercap_1952` (since it's a column name)

## Select multiple columns or rows using `DataFrame.iloc` and a named slice.

In [35]:
print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'])

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy           8243.582340    10022.401310    12269.273780
Montenegro      4649.593785     5907.850937     7778.414017
Netherlands    12790.849560    15363.251360    18794.745670
Norway         13450.401510    16361.876470    18965.055510
Poland          5338.752143     6557.152776     8006.506993


In the above code, we discover that **slicing using indexes is inclusive at both ends**, which differs from typical python behavior where slicing indicates everything up to but not including the final index.

However, if we use integers when our DataFrame is indexed by something else, slicing follows typical pythonic behavior.

## Result of slicing can be used in further operations.

*   Usually don't just print a slice.
*   All the statistical operators that work on entire data frames
    work the same way on slices.

In [78]:
print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].max())

gdpPercap_1962    13450.40151
gdpPercap_1967    16361.87647
gdpPercap_1972    18965.05551
dtype: float64


In [79]:
print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].min())

gdpPercap_1962    4649.593785
gdpPercap_1967    5907.850937
gdpPercap_1972    7778.414017
dtype: float64


## Use comparisons to select data based on value.

*   Comparison is applied element by element.
*   Returns a similarly-shaped data frame of `True` and `False`.

In [80]:
# Use a subset of data to keep output readable.
subset = data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972']
print('Subset of data:\n', subset)

# Which values were greater than 10000 ?
print('\nWhere are values large?\n', subset > 10000)

Subset of data:
              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy           8243.582340    10022.401310    12269.273780
Montenegro      4649.593785     5907.850937     7778.414017
Netherlands    12790.849560    15363.251360    18794.745670
Norway         13450.401510    16361.876470    18965.055510
Poland          5338.752143     6557.152776     8006.506993

Where are values large?
              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy                 False            True            True
Montenegro            False           False           False
Netherlands            True            True            True
Norway                 True            True            True
Poland                False           False           False


## Select values or NaN using a Boolean mask.

*   A frame full of Booleans is sometimes called a *mask* because of how it can be used.

In [81]:
mask = subset > 10000
print(subset[mask])

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy                   NaN     10022.40131     12269.27378
Montenegro              NaN             NaN             NaN
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551
Poland                  NaN             NaN             NaN


*   Get the value where the mask is true, and NaN (Not a Number) where it is false.
*   Useful because NaNs are ignored by operations like max, min, average, etc.

In [82]:
print(subset[subset > 10000].describe())

       gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
count        2.000000        3.000000        3.000000
mean     13120.625535    13915.843047    16676.358320
std        466.373656     3408.589070     3817.597015
min      12790.849560    10022.401310    12269.273780
25%      12955.737547    12692.826335    15532.009725
50%      13120.625535    15363.251360    18794.745670
75%      13285.513523    15862.563915    18879.900590
max      13450.401510    16361.876470    18965.055510


# Exercises

> ## Selection of Individual Values
>
> Assume Pandas has been imported into your notebook and the Gapminder GDP data for Europe has been loaded:
>
> ~~~
> import pandas
>
> df = pandas.read_csv('data/gapminder_gdp_europe.csv', index_col='country')
> ~~~
>
> Write an expression to find the Per Capita GDP of Serbia in 2007.

>> ### Solution:
>>The selection can be done by using the labels for both the row (“Serbia”) and the column (“gdpPercap_2007”):

>>`print(df.loc['Serbia', 'gdpPercap_2007'])`

>>The output is

>>`9786.534714`

> ## Extent of Slicing
>
> 1.  Do the two statements below produce the same output?
> 2.  Based on this,what rule governs what is included (or not) in numerical slices and named slices in Pandas?
>
> ~~~
> print(data.iloc[0:2, 0:2])
> print(data.loc['Albania':'Belgium', 'gdpPercap_1952':'gdpPercap_1962'])
> ~~~
> 
>>### Solution
>> No, they do not produce the same output! The output of the first statement is:

>>        gdpPercap_1952  gdpPercap_1957
>>country                                
>>Albania     1601.056136     1942.284244
>>Austria     6137.076492     8842.598030

>>The second statement gives:
~~~
>>           gdpPercap_1952  gdpPercap_1957  gdpPercap_1962 \
>>country                                                
>>Albania     1601.056136     1942.284244     2312.888958
>>Austria     6137.076492     8842.598030    10750.721110
>>Belgium     8343.105127     9714.960623    10991.206760
~~~
>>Clearly, the second statement produces an additional column compared to the first statement. What conclusion can we draw? We see that a numerical slice, 0:2, omits the final index (i.e. index 2) in the range provided, while a named slice, ‘gdpPercap_1952’:’gdpPercap_1962’, includes the final element

> ## Reconstructing Data
>
> Explain what each line in the following short program does:
> what is in `first`, `second`, etc.?
>
> ~~~
> first = pandas.read_csv('data/gapminder_gdp_all.csv', index_col='country')
> second = df[df['continent'] == 'Americas']
> third = second.drop('Puerto Rico')
> fourth = third.drop('continent', axis = 1)
> fourth.to_csv('result.csv')
> ~~~

>>### Solution
>>Let’s go through this piece of code line by line.
>>
>>`first = pandas.read_csv('data/gapminder_all.csv', index_col='country')`
>>
>>This line loads the dataset containing the GDP data from all countries into a dataframe called first. The index_col='country' parameter selects which column to use as the row labels in the dataframe.
>>
>>`second = first[first['continent'] == 'Americas']`
>>
>>This line makes a selection: only those rows of first for which the ‘continent’ column matches ‘Americas’ are extracted. Notice how the Boolean expression inside the brackets, first['continent'] == 'Americas', is used to select only those rows where the expression is true. Try printing this expression! Can you print also its individual True/False elements? (hint: first assign the expression to a variable)
>>
>>`third = second.drop('Puerto Rico')`
>>
>>As the syntax suggests, this line drops the row from second where the label is ‘Puerto Rico’. The resulting dataframe third has one row less than the original dataframe second.
>>
>>`fourth = third.drop('continent', axis = 1)`
>>
>>Again we apply the drop function, but in this case we are dropping not a row but a whole column. To accomplish this, we need to specify also the axis parameter (we want to drop the second column which has index 1).
>>
>>`fourth.to_csv('result.csv')`
>>
>>The final step is to write the data that we have been working on to a csv file. Pandas makes this easy with the to_csv() function. The only required argument to the function is the filename. Note that the file will be written in the directory from which you started the Jupyter or Python session.

> ## Selecting Indices
>
> Explain in simple terms what `idxmin` and `idxmax` do in the short program below.
> When would you use these methods?
>
> ~~~
> data = pandas.load_csv('data/gapminder_gdp_europe.csv', index_col='country')
> print(data.idxmin())
> print(data.idxmax())
> ~~~
>> ### Solution
>> For each column in data, idxmin will return the index value corresponding to each column’s minimum; idxmax will do accordingly the same for each column’s maximum value.

> ## Practice with Selection.
>
> Assume Pandas has been imported and the Gapminder GDP data for Europe has been loaded.
> Write an expression to select each of the following:
>
> 1.  GDP per capita for all countries in 1982.
> 2.  GDP per capita for Denmark for all years.
> 3.  GDP per capita for all countries for years *after* 1985.
> 4.  GDP per capita for each country in 2007 as a multiple of 
>     GDP per capita for that country in 1952.

>>### Solution
>>1: `data['gdpPercap_1982']`
>>
>>2: `data.loc['Denmark',:]`
>>
>>3: `data.loc[:,'gdpPercap_1985':]`
>>
>>4: `data['gdpPercap_2007']/data['gdpPercap_1952']`
>>

>> ## Interpretation
>>
>> Poland's borders have been stable since 1945, but changed several times in the years before then. How would you handle this if you were creating a table of GDP per capita for Poland for the entire Twentieth Century?

---
## Keypoints:
 - "Use `DataFrame.iloc[..., ...]` to select values by index location."
 - "Use `:` on its own to mean all columns or all rows."
 - "Select multiple columns or rows using `DataFrame.ix` and a named slice."
 - "Result of slicing can be used in further operations."
 - "Use comparisons to select data based on value."
 - "Select values or NaN using a Boolean mask."